xdsl/backend/register_queue.py

```diff
@@ -64,3 +66,143 @@ def exclude_register(self, reg: _T) -> None:
         Removes regiesters from available set, if present.
         """
         ...
+
+
+@dataclass
+class LIFORegisterQueue(Generic[_T], RegisterQueue[_T]):
+    """
+    LIFO queue of registers available for allocation.
+    """
+
+    next_infinite_indices: defaultdict[str, int] = field(
+        default_factory=lambda: defaultdict(lambda: 0)
+    )
+    """Next index for a given register set."""
+
+    reserved_registers: defaultdict[str, defaultdict[int, int]] = field(
+        default_factory=lambda: defaultdict(lambda: defaultdict[int, int](lambda: 0))
+    )
+    """
+    Registers unavailable to be used by the register allocator for a given register set.
+    """
+
+    available_registers: defaultdict[str, list[int]] = field(
+        default_factory=lambda: defaultdict(list[int])
+    )
+    """
+    Registers from a given register set that values can be allocated to in the current
+    context.
+    """
+
+    @classmethod
+    def default_reserved_registers(cls) -> Iterable[_T]:
+        """
+        The default registers to be made unavailable when instantiating the queue.
+        """
+        return ()
+
+    @classmethod
+    def default_available_registers(cls) -> Iterable[_T]:
+        """
+        The default registers to be made available when instantiating the queue.
+        """
+        return ()
+
+    @classmethod
```

Comment:
from here and below, the code is a copy/paste from `RiscvRegisterQueue`, changing `RISCVRegisterType` to `_T`